multiversx_sdk/abi/abi_definition.py

```diff
@@ -0,0 +1,260 @@
+                 types: "TypesDefinitions") -> None:
+        self.constructor = constructor
+        self.upgrade_constructor = upgrade_constructor
+        self.endpoints = endpoints
+        self.types = types
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "AbiDefinition":
+        constructor = EndpointDefinition.from_dict(data["constructor"])
+        constructor.name = "constructor"
+
+        upgrade_constructor = cls._get_endpoint_for_upgrade(data)
+        upgrade_constructor.name = "upgrade_constructor"
+
+        endpoints = [EndpointDefinition.from_dict(item) for item in data["endpoints"]] if "endpoints" in data else []
+        types = TypesDefinitions.from_dict(data.get("types", {}))
+
+        return cls(
+            constructor=constructor,
+            upgrade_constructor=upgrade_constructor,
+            endpoints=endpoints,
+            types=types
+        )
+
+    @classmethod
+    def _get_endpoint_for_upgrade(cls, data: Dict[str, Any]) -> "EndpointDefinition":
+        if "upgradeConstructor" in data:
+            return EndpointDefinition.from_dict(data["upgradeConstructor"])
+
+        # Fallback for contracts written using a not-old, but not-new Rust framework:
+        if "upgrade" in data["endpoints"]:
+            return EndpointDefinition.from_dict(data["endpoints"]["upgrade"])
+
+        # Fallback for contracts written using an old Rust framework:
+        return EndpointDefinition.from_dict(data["constructor"])
+
+    @classmethod
+    def load(cls, path: Path) -> "AbiDefinition":
+        content = Path(path).read_text()
+        data = json.loads(content)
+        return cls.from_dict(data)
+
+
+class EndpointDefinition:
+    def __init__(self,
+                 name: str,
+                 mutability: str,
+                 inputs: List["ParameterDefinition"],
+                 outputs: List["ParameterDefinition"],
+                 payable_in_tokens: List[str],
+                 only_owner: bool) -> None:
+        self.name = name
+        self.mutability = mutability
+        self.inputs = inputs
+        self.outputs = outputs
+        self.payable_in_tokens = payable_in_tokens
+        self.only_owner = only_owner
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "EndpointDefinition":
+        inputs = [ParameterDefinition.from_dict(item) for item in data["inputs"]]
+        outputs = [ParameterDefinition.from_dict(item) for item in data["outputs"]]
+
+        return cls(
+            name=data.get("name", ""),
+            mutability=data.get("mutability", ""),
+            inputs=inputs,
+            outputs=outputs,
+            payable_in_tokens=data.get("payableInTokens", []),
+            only_owner=data.get("onlyOwner", False)
+        )
+
+    def __repr__(self):
+        return f"EndpointDefinition(name={self.name})"
+
+
+class ParameterDefinition:
+    def __init__(self, name: str, type: str) -> None:
+        self.name = name
+        self.type = type
```

Comment:
Maybe (I don't know if those are relevant) `multi_arg: bool` for inputs or `multi_result: bool` for outputs

Reply:
They are redundant / not relevant, indeed. Sufficient information is captured in the type expression itself.